pkg/codegen/dotnet/gen.go

```diff
@@ -1081,6 +1143,70 @@ func (mod *modContext) genResource(w io.Writer, r *schema.Resource) error {
 		}
 	}
 
+	// Generate method types.
+	genMethodTypes := func(method *schema.Method) error {
```

Comment:
I was originally planning to emit the `<Method>Args` and `<Method>Result` classes as nested classes within the resource's class, e.g.:

```csharp
public class MyResource : ComponentResource
{
    public Output<FooResult> Foo(FooArgs args)
        => ...

    public sealed class FooArgs : Pulumi.CallArgs
    {
        ...
    }

    public sealed class FooResult
    {
    }
}
```

Which is similar to what we've done for Node.js and Python, and appropriately "namespaces" the types to avoid any naming conflicts for similarly named methods on different resources within the same Pulumi module.

However, using these as nested classes from .NET is more annoying which isn't an issue in practice with TS/JS and Python. In .NET you have to do:

```
var res = new MyResource(...);
var result = res.Foo(MyResource.FooArgs(...)); // Kinda annoying
```

When it'd be much nicer to use as:

```
var res = new MyResource(...);
var result = res.Foo(FooArgs(...)); // Nicer
```

This PR currently does the latter nicer approach (non-nested).

**What do folks think?** 

There is some risk that within a Pulumi module, there could be resources that have methods with the same name which then would result in conflicts. If that becomes a problem, we'd have to come up with a way to address it, either by nesting the classes, prefixing the resource name (e.g. `<Resource><Method>Args`), or providing a way to customize the naming from the schema. However, I wonder how common such conflicts would be in practice (same method name on multiple resources in the same module).

Note that in Go, we currently emit these prefixed as `<Resource><Method>Args` and `<Resource><Method>Result`, to prevent such conflicts, but this is also annoying to use, so if we proceed with the unnested, un-resource-prefixed names in C#, I'll open a PR to remove the `<Resource>` prefix from the Go codegen.

Reply:
On the other hand, there's already precedent for prefixing the resource name for nested types, e.g. `Bucket`'s `Website` property accepts a `BucketWebsiteArgs` type. This is making me lean towards making the args/result types non-nested and prefixed with the resource name, e.g.:

```csharp
var res = new MyResource(...);
var result = res.Foo(new MyResourceFooArgs(...));
```

It isn't as nice to use, but it's more consistent with what we've done elsewhere and reduces the likelihood of conflicts.

Reply:
Flattening them out into method params without an intermediate class is out of question? After all these are methods, so one expects to call them without much ceremony? Most of these will have 1-2 parameters?

Reply:
> Flattening them out into method params without an intermediate class is out of question? After all these are methods, so one expects to call them without much ceremony? Most of these will have 1-2 parameters?

It's a great question. From a usage standpoint, I'd love for these to be params. I think the main downside is we don't provide much control over the order of these as specified in the schema, which makes adding new optional params in a compatible way problematic. I believe this is why we've always emitted args classes for functions, rather than flattened into params -- so what we're doing for methods is just consistent with the approach we've taken for functions.

Note that we do flatten these to method args in Python only, but force the args to be specified by keyword rather than positionally, so if new args are added, existing code continues to work. For .NET, there's no way to force params to be specified by name, though.

For methods, we could decide to be different from what we've done for functions, and flatten them out into method params, and just live with the limited versionablity, i.e. if you want to add a new optional arg in a compatible way, you'd have to add a new method instead of adding it to an existing method.

Reply:
I think consistency of methods and functions is important to keep, as well as consistency across languages. I filed the issue citing your excellent notes in https://github.com/pulumi/pulumi/issues/7664 - it probably has no legs but just in case we one day find a way forward on this, I wanted to keep a note in the tracker.